### PR TITLE
Collision detection for PBC in DEM and CFD-DEM

### PR DESCRIPTION
**Pull request by acdaigneault**
# Description of the problem

- Mapping of periodic boundaries and particle displacement was added in the PR #557 
- This previous PR did not consider collision detection of particles in periodic boundaries

# Description of the solution

- Enabling the collision detection for particles in the cells of the periodic boundaries
- The particle containers and the way it is implemented is really similar to the way current collision detection are (broad search, updating of candidate, fine search, force calculation, ...). But it needed proper containers since we need sort of tracking of the particles that has periodic candidates in periodic boundary for the neighbor threshold search and the displacement with offset.
- 3 new types of particle collisions : periodic local-local collisions,  periodic local-ghost collisions & periodic ghost-local collisions.
- One direction periodic mapping : periodic cells or periodic collisions are only mapped related to the cells/particles on the periodic boundary 0 (this is why there's local-ghost & ghost local containers)
- It is not possible to use more than one pair of periodic boundaries now

# How Has This Been Tested?

- Add the `periodic_boundary_collisions` application test for DEM where 2 particles collide in the periodic boundaries (mpirun=1 & mpirun=2)
- `periodic_multiple_boundaries` is removed since we don't support more than 1 pair of periodic boundaries
- All other tests worked

# Future changes

- Need some clean up in contact forces code with ContactType to remove recurring code in future PR
- Find a cleaner way to get offset
- Add some content to the DEM hopper example to show the periodic boundary options

### CODE REVIEW

File: include/dem/find_cell_neighbors.h [hunk 1]
```diff
@@ -77,15 +108,41 @@ class FindCellNeighbors
...
 
   void
   find_full_cell_neighbors(
     const parallel::distributed::Triangulation<dim> &triangulation,
     typename DEM::dem_data_structures<dim>::cells_total_neighbor_list
       &cells_total_neighbor_list);
+
+private:
+  /**
+   * Generate a periodic neighbor cells list of the cell.
```
Comment (blaisb) [inline include/dem/find_cell_neighbors.h:123]: I think this could be a bit more verbose. Can you give an example of what that contains?

File: include/dem/find_cell_neighbors.h [hunk 2]
```diff
@@ -69,6 +69,37 @@ class FindCellNeighbors
     typename DEM::dem_data_structures<dim>::cells_neighbor_list
       &cells_ghost_neighbor_list);
 
+  /**
+   * Finds the neighbor list (without repetition) of all the active cells in the
+   * triangulation
```
Comment (blaisb) [inline include/dem/find_cell_neighbors.h:74]: I would use this occasion to make this brief a bit more verbose. Also please use the @brief keyword.

File: include/dem/find_cell_neighbors.h (hunk 1 above)
Comment (blaisb) [reply include/dem/find_cell_neighbors.h:123]: Also @brief keyword

File: include/dem/particle_particle_broad_search.h [hunk 3]
```diff
@@ -57,16 +57,34 @@ class ParticleParticleBroadSearch
...
   void
   find_particle_particle_contact_pairs(
     dealii::Particles::ParticleHandler<dim> &particle_handler,
     DEMContainerManager<dim> &               container_manager);
 
+
+  /**
+   * Finds a vector of pairs (particle_particle_candidates) which shows the
+   * candidate particle-particle collision pairs. These collision pairs will be
+   * used in the fine search to investigate if they are in contact or not.
```
Comment (blaisb) [inline include/dem/particle_particle_broad_search.h:73]:
```suggestion
   * @brief Finds a vector of pairs (particle_particle_candidates) which contains the
   * candidate particle-particle collision pairs. These collision pairs will be
   * used in the fine search to investigate if they are in contact or not.
```

File: include/dem/periodic_boundaries_manipulator.h [hunk 4]
```diff
@@ -32,12 +33,8 @@ using namespace dealii;
 #  define particle_wall_periodic_displacement_h
 
 /**
- * This class is used to manipulate the particles location when they cross a
- * periodic boundary
- *
- * @note Particle collisions across periodic boundaries are currently not
- * implemented.
-
+ * This class is used to manipulate the particles location when they cross
```
Comment (blaisb) [inline include/dem/periodic_boundaries_manipulator.h:36]: Can you be a bit more verbose. Explain what manipulating means here.

File: include/dem/periodic_boundaries_manipulator.h [hunk 5]
```diff
@@ -81,15 +82,29 @@ class PeriodicBoundariesManipulator
...
   void
   execute_particles_displacement(
-    const Particles::ParticleHandler<dim> &particle_handler);
+    const Particles::ParticleHandler<dim> &particle_handler,
+    typename DEM::dem_data_structures<dim>::periodic_boundaries_cells_info
+      &periodic_boundaries_cells_information);
+
+  /**
+   * @brief Return the periodic offset, it is get from the first pair of cells
+   * on periodic boundaries, all pair of cells have the same offset
```
Comment (blaisb) [inline include/dem/periodic_boundaries_manipulator.h:96]:
```suggestion
   * @brief Return the periodic offset, it is calculated from the first pair of cells
   * on periodic boundaries, all pair of cells are assumed to have the same offset
```

File: source/dem/particle_particle_contact_force.cc [hunk 6]
```diff
@@ -469,6 +479,518 @@ ParticleParticleContactForce<dim, contact_model>::
...
+                {
+                  // if the adjacent pair is not in contact anymore, only the
+                  // tangential overlap is set to zero
+                  contact_info.tangential_overlap.clear();
+                }
+            }
+        }
+    }
+
+  for (auto &&periodic_adjacent_particles_list :
```
Comment (blaisb) [inline source/dem/particle_particle_contact_force.cc:830]: Yeah ok clearly refactoring these loops will be a good thing to do in a next PR.

**Review by blaisb — CHANGES_REQUESTED** [review]
Just a few places where you can improve the documentation.
The refactoring of the force calculation will be on my ballpark in the following weeks.
Can you also update the documentation accordingly to your changes? It would be a good thing to also add a small section to the hopper 2D example where you launch it again in periodic like you tried, this way the feature would be already demonstrated somewhere. Either in this PR or in a quick next PR, as you see fit.

Obviously the tests also need to be fixed

File: include/dem/find_cell_neighbors.h (hunk 1 above)
Comment (acdaigneault) [reply include/dem/find_cell_neighbors.h:123]: Do you know why sometimes there's no @brief? I still have the doc in CLion, with or without

**Comment by acdaigneault** [comment]
@blaisb 
- [x] Fix code for DEM tests
- [x] Add basic documentation in the Lethe wiki for PBC
- [x] Improve description of functions in code
- [ ] Modify 3d hopper example (will be in another PR)

File: doc/source/parameters/dem/boundary_conditions.rst [hunk 7]
```diff
@@ -49,9 +49,16 @@ In this subsection, the boundary conditions of the DEM simulation are defined. F
...
 * For each boundary condition, we have to define a separate subsection. In the sample parameter list above, the ``number of boundary conditions`` is equal to 2. Hence, we need to define two subsections (``subsection boundary condition 0`` and ``subsection boundary condition 1``).
 
-* The ``boundary id`` parameter specifies the boundary ID for which the boundary condition should be applied.
+* The ``boundary id`` parameter specifies the boundary ID for which the boundary condition should be applied, periodic boundaries are an exception.
 
-* The ``type`` parameter specifies the type of the boundary condition. Acceptable types are: ``fixed_wall``, ``outlet``, ``rotational``, and ``translational``. The default boundary condition type is ``fixed_wall``.
+* The ``periodic id 0`` and ``periodic id 1`` parameters specify the periodic boundaries ID for which the periodic boundary condition should be applied.
+
+.. note::
+    Only parallel periodic walls are currently working.
```
Comment (blaisb) [inline doc/source/parameters/dem/boundary_conditions.rst:57]:
```suggestion
    Only periodic boundaries which have co-linear normal vectors which align along one axis of the problem (e.g., x axis) are currently supported. 
```

File: include/dem/periodic_boundaries_manipulator.h [hunk 8]
```diff
@@ -28,16 +29,19 @@
...
+ * and manipulate the particles location when they cross periodic boundaries,
+ * i.e. when a particle cross a periodic boundary (0 or 1) its location is
+ * modified with the offset between faces of periodic cells on periodic
+ * boundaries.
  *
- * @note Particle collisions across periodic boundaries are currently not
- * implemented.
-
+ * @note Currently the code only supports one pair of periodic boundaries and
+ * those boundaries must be parallel.
```
Comment (blaisb) [inline include/dem/periodic_boundaries_manipulator.h:44]:
```suggestion
 * @note Currently the code only supports one pair of periodic boundaries and
 * those boundaries must be parallel and aligned with an axis of the domain (e.g., x axis).
```

**Review by blaisb — CHANGES_REQUESTED** [review]
Two minor wording change then this is ready for a merge. Good job!!!!! Very good job!!!

**Review by blaisb — APPROVED** [review]
